### PR TITLE
Functions: Run project against a function core tool version defined in proj file #294

### LINKED ISSUE #294: Functions: Run project against a function core tool version defined in proj file
Original request - https://github.com/JetBrains/azure-tools-for-intellij/issues/278

Project file can specify multiple functions with different core tools version:
```
<PropertyGroup>
    <AzureFunctionsVersion>V3</AzureFunctionsVersion>
</PropertyGroup>
```
We need to define a plugin behaviour for this case and check if we need to run against a specified version.

### PR DESCRIPTION
**Pull request by maartenba**
Fixes #294

Do not merge yet:
* [x] Needs  #557 to be merged
* [x] Need to be rebased after review
* [x] Also need to double-check https://github.com/JetBrains/azure-tools-for-intellij/pull/557#pullrequestreview-860863644 after all rebases.
* [x] Needs changelog updates

# Functionality

With this PR, the Azure Functions Core Tools can now be configured per version (`v2`, `v3`,  and `v4`). Depending on the version of the `AzureFunctionsVersion` MSBuild variable, Rider will pick the tooling for that version.

## Settings

The settings screen has been reworked to show a table of versions and their tools path mapping, and an option to specify where tools should be downloaded automatically.

The value for path can be:
* (empty), which will show `Managed by Rider` - Rider will manage tools for you and download them for you
* An absolute path - Rider will use this path for the specific tools version
* `func` - Rider will use the `func`/`func.exe` found in your environment, e.g. the one installed with Chocolatey, NPM or Homebrew

settings-screen

The download path defaults to `$HOME/.AzureToolsForIntelliJ/AzureFunctionsCoreTools` (we are already using `$HOME/.AzureToolsForIntelliJ` to store authentication token caches etc.)

## Automatic downloads

For every tools version, the default is `Managed by Rider`. Rider will manage tools for you and download them from the feed at `https://functionscdn.azureedge.net/public/cli-feed-v4.json` (used by VS2022 as well).

* Download happens just before running/debugging your application. Since the MSBuild variable may change at any time, this seems like the best moment to do this.
* Once per Rider instance, this will also check for updates. If a newer version of the tools is available, Rider will update them

Here's this feature in action:

download-tools-on-run

## Project template

Project templates use the latest templates found. Right now, this means `v4` templates are used. I have added an initial configuration option. before, you could specify the path to the core tools and Rider would detect templates, in this new version you can download the tools immediately:

project-templates

# Registry flags

Two new registry flags were added:
* `azure.function_app.core_tools.feed.enabled` - Enable/disable the automatic download behaviour. Enabled by default.
* `azure.function_app.core_tools.feed.url` - URL to the core tools feed, can be updated to a URL in your intranet if you're in a firewalled environment. Defaults to `https://functionscdn.azureedge.net/public/cli-feed-v4.json`

# Migration

A startup activity `CleanupDeprecatedAzureCoreToolsConfigurationActivity` has been added that will remove the old "managed by Rider" directory. This will save the user 100-200MB of disk space depending on what version of the tools they are currently using. We will need to remove this startup activity at some point in the future.

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/test/kotlin/org/jetbrains/plugins/azure/functions/helpers/csharp/AzureFunctionsPsiHelperTest.kt [hunk 1]
```diff
@@ -25,6 +25,7 @@ package org.jetbrains.plugins.azure.functions.helpers.csharp
 import com.jetbrains.rider.test.asserts.shouldBeFalse
 import org.testng.annotations.Test
 
+// REVIEW @sdubov: Integration tests exist for TimerTrigger completion etc. Should this class be removed?
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/test/kotlin/org/jetbrains/plugins/azure/functions/helpers/csharp/AzureFunctionsPsiHelperTest.kt:28]: FYI, removed this test in the other PRs as we indeed have an integration test

**Comment by sdubov** [comment]
@maartenba, thank you a lot for this detailed description of the feature! Couple hi-level notes from my side: 

**Settings:**
1. How do a user know about possible options? e.g. `Empty` and `full path` are pretty clear, but I'm not sure `func`/`func.exe` options are discoverable without any hints.
2. I like the new UI. I would probably post those images into out internal channels to discuss the UI/UX as we usually do since I have number of concern: 
   1) I think we have pretty huge table for 3 rows. They are pretty much constant and is not extendable from a user side. I would make this table wrap contend vertically. What do you think?
CoreToolVersion
   2) I think the layout for "Azure Functions Core tool download path:" is a bit messy. I'm not sure how to make it better, but maybe split by two lines?
CoreToolPath
   3) Please feel free to ignore these comments if you find your solution works better.


3. Automatic download is applied for `Managed by Rider` option only, right?


**Project templates:**
1. I stuck on a "configure Azure Core Tool..." state (only Configure option is available) if I have a template (probably, previously installed), but my Azure Core Tool is not configured. So, I click "Configure Azure Core Tool" and I see each of them is "Managed by Rider" (expected), I click Ok and do not see a template because Core tool is not downloaded yet. I think we need to handle an exact this case and download necessary Azure Core Tool if it does not exist after we back to the templates window.

**Registry flags:**
1. What do you think about moving those flags on a Settings page and add UI? So, a user can easily configure them. 

**Migration:**
1. What do you think if we show a prompt with a description why we are doing that and to get a confirmation about this option? Deleting something on a user box sounds a bit scary :)

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/AzureRiderSettings.kt [hunk 2]
```diff
@@ -40,9 +44,47 @@ object AzureRiderSettings {
     const val OPEN_IN_BROWSER_AFTER_PUBLISH_DEFAULT_VALUE = false
 
     // Functions
-    const val PROPERTY_FUNCTIONS_CORETOOLS_PATH = "AzureFunctionsCoreToolsPath"
-    const val PROPERTY_FUNCTIONS_CORETOOLS_ALLOW_PRERELEASE = "AzureFunctionsCoreToolsAllowPrerelease"
-    const val PROPERTY_FUNCTIONS_CORETOOLS_CHECK_UPDATES = "AzureFunctionCoreToolsCheckUpdates"
+    const val PROPERTY_FUNCTIONS_CORETOOLS_PATHS = "AzureFunctionsCoreToolsPaths"
+    const val PROPERTY_FUNCTIONS_CORETOOLS_DOWNLOAD_PATH = "AzureFunctionsCoreToolsDownloadPath"
+    val VALUE_FUNCTIONS_CORETOOLS_DOWNLOAD_PATH: String = PathUtil.toSystemIndependentName(
+            File(System.getProperty("user.home"), AZURE_TOOLS_FOLDER).resolve("AzureFunctionsCoreTools").absolutePath)
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/AzureRiderSettings.kt]: JFYI, there is a helper method `FileUtil.getDirectoryWithinUserHome()`.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/AzureRiderSettings.kt [hunk 3]
```diff
@@ -40,9 +44,47 @@ object AzureRiderSettings {
...
+    }
+
+    fun getAzureCoreToolsPathEntries(properties: PropertiesComponent) : List<AzureCoreToolsPathEntry> {
+
+        val coreToolsFromConfiguration = properties.getValues(PROPERTY_FUNCTIONS_CORETOOLS_PATHS)
+                ?.mapNotNull { AzureCoreToolsPathEntry.fromStringEntry(it) }
+                ?: emptyList()
+
+        val coreTools = FunctionsCoreToolsConstants.FUNCTIONS_CORETOOLS_KNOWN_SUPPORTED_VERSIONS.map { functionsVersion ->
+            coreToolsFromConfiguration.firstOrNull { it.functionsVersion == functionsVersion }
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/AzureRiderSettings.kt]: `ignoreCase = true` - ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/ui/AzureFunctionsConfigurationPanel.kt [hunk 4]
```diff
@@ -22,168 +22,152 @@
...
-        private const val CARD_BUTTON = "button"
-        private const val CARD_PROGRESS = "progress"
-    }
+    private lateinit var coreToolsEditorModel: ListTableModel<AzureRiderSettings.AzureCoreToolsPathEntry>
+    private lateinit var coreToolsEditor: JBTable
+    private lateinit var coreToolsDownloadPathEditor: TextFieldWithBrowseButton
 
-    private val unknownLabel: String = "<${message("common.unknown_lower_case")}>"
+    private val coreToolsEditorColumns = arrayOf<ColumnInfo<*, *>>(
+            object : ColumnInfo<AzureRiderSettings.AzureCoreToolsPathEntry, String>(message("settings.app_services.function_app.core_tools.configuration.column.functionsVersion")) {
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/ui/AzureFunctionsConfigurationPanel.kt:72]: I think we can change layout for columns from 50/50 to e.g. 30/70 to fully fit the "Azure Function Version" title. "v2(3,4)" does not take too much place, while the path could take more place.

**Comment by maartenba** [comment]
Thanks, if it's okay for you @sdubov I will start working on those after I get a thumbs up on #557 and #558 if that's okay (so I don't have to drag along 3 PRs and keep them synced)

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/CleanupDeprecatedAzureCoreToolsConfigurationActivity.kt [hunk 5]
```diff
@@ -23,39 +23,39 @@
...
-            logger.info("Azure Function Core Tool is setup: '$existingCoreToolPath'")
-            return
+        @Suppress("DEPRECATION")
+        val logger = Logger.getInstance(CleanupDeprecatedAzureCoreToolsConfigurationActivity::class.java)
+
+        // Remove old Azure Core Tools directory
+        val coreToolsDirectory = File(PathManager.getConfigPath()).resolve("azure-functions-coretools")
+        if (coreToolsDirectory.exists()) {
+            try {
+                coreToolsDirectory.deleteRecursively()
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/CleanupDeprecatedAzureCoreToolsConfigurationActivity.kt]: Ask user's confirmation? I concern a bit about the case when we silently remove some tools that a user might already use. What do you think?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsInfoProvider.kt [hunk 6]
```diff
@@ -24,28 +24,95 @@ package org.jetbrains.plugins.azure.functions.coreTools
...
+
+        // If configuration is func/func.cmd/func.exe, try and determine the full path from the environment
+        if (coreToolsPathFromConfiguration.equals("func", ignoreCase = true) ||
+                coreToolsPathFromConfiguration.equals("func.cmd", ignoreCase = true) ||
+                coreToolsPathFromConfiguration.equals("func.exe", ignoreCase = true)) {
+
+            val coreToolsInfoFromEnvironment = FunctionCliResolver.resolveFunc()
+                    ?.let { resolveFromPath(File(it)) }
+
+            return coreToolsInfoFromEnvironment // can be null: if func.exe is configured but not found, user needs to check settings
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsInfoProvider.kt:85]: Do we want to notify a user about this case? From code it looks like a user will never know about func/func.cmd/func.exe was not found. Maybe at least log this info?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsMsBuild.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    const val PROPERTY_AZURE_FUNCTIONS_VERSION = "AzureFunctionsVersion"
+
+    fun requestAzureFunctionsVersion(project: Project, projectFilePath: String): String? {
+
+        val msBuildEvaluator = MSBuildEvaluator.getInstance(project)
+
+        val msBuildPropertiesPromise = msBuildEvaluator.evaluateProperties(
+                MSBuildEvaluator.PropertyRequest(projectFilePath, null, listOf(PROPERTY_AZURE_FUNCTIONS_VERSION)))
+
+        // REVIEW: Should this be shorter?
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsMsBuild.kt]: Why is 1 minute? I would say it should be quite fast operation to process msbuild properties.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/test/kotlin/org/jetbrains/plugins/azure/functions/helpers/csharp/AzureFunctionsPsiHelperTest.kt (hunk 1 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/test/kotlin/org/jetbrains/plugins/azure/functions/helpers/csharp/AzureFunctionsPsiHelperTest.kt:28]: @maartenba, please drop this class. I'm not sure what this test should check currently.

**Comment by sdubov** [comment]
@maartenba, sure. I'll do the rest of reviews in a row.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsMsBuild.kt (hunk 7 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsMsBuild.kt]: Just took a number. Any suggestions on a better timeout? E.g. 10 seconds or so? (I remember @xtmq at one point seeing it's not lightning fast)

**Comment by maartenba** [comment]
[ ] Also need to double-check https://github.com/JetBrains/azure-tools-for-intellij/pull/557#pullrequestreview-860863644 after all rebases.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsMsBuild.kt (hunk 7 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsMsBuild.kt]: Switched to 15 sec.

**Comment by maartenba** [comment]
@sdubov

UI updates:
* Table was made smaller, and column sizes were updated as well
* Azure Functions Core tool download path selector is now on its own line in the UI, making it wider

image

To answer your questions:
* Automatic download is applied for `Managed by Rider` option only indeed.
* _I stuck on a "configure Azure Core Tool..." state (only Configure option is available)_ - should be fixed now
* Registry entries - keeping them for now. Changing feed URL will normally not be needed. And folks can always specify paths manually.
* Migration - Now shows a notification instead, so it requires user interaction and we're not just deleting from filesystem

image

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsInfoProvider.kt (hunk 6 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsInfoProvider.kt:85]: Added `warn` logging

**Comment by maartenba** [comment]
Just pushed some settings UI updates. What do you think, @sdubov ? (dropdown location is because of gif recording)

Untitled Project

**Review by sdubov — APPROVED** [review]

**Comment by sdubov** [comment]
@maartenba, great work! I like all new changes for Core tools.

**Pull request merged**